arbor/include/arbor/simd/simd.hpp

```diff
@@ -17,41 +18,336 @@ namespace detail {
 
     template <typename Impl>
     struct simd_mask_impl;
+
+    template <typename To>
+    struct simd_cast_impl;
+
+    template <typename I, typename V>
+    class indirect_indexed_expression;
+
+    template <typename V>
+    class indirect_expression;
+
+    template <typename T, typename M>
+    class where_expression;
+
+    template <typename T, typename M>
+    class const_where_expression;
+}
+
+// Top level functions for second API
+using detail::simd_impl;
+using detail::simd_mask_impl;
+
+template <typename Impl>
+inline constexpr int width(const simd_impl<Impl>& a) {
+    return simd_impl<Impl>::width;
+};
+
+template <typename Impl>
+typename simd_impl<Impl>::scalar_type sum(const simd_impl<Impl>& a) {
+    return a.sum();
+};
+
+#define UNARY_ARITHMETIC(name)\
+template <typename Impl>\
+simd_impl<Impl> name(const simd_impl<Impl>& a) {\
+    return simd_impl<Impl>::wrap(Impl::name(a.value_));\
+};
+
+#define BINARY_ARITHMETIC(name)\
+template <typename Impl>\
+simd_impl<Impl> name(const simd_impl<Impl>& a, simd_impl<Impl> b) {\
+    return simd_impl<Impl>::wrap(Impl::name(a.value_, b.value_));\
+};\
+template <typename Impl>\
+simd_impl<Impl> name(const simd_impl<Impl>& a, typename simd_impl<Impl>::scalar_type b) {\
+    return simd_impl<Impl>::wrap(Impl::name(a.value_, Impl::broadcast(b)));\
+};\
+template <typename Impl>\
+simd_impl<Impl> name(const typename simd_impl<Impl>::scalar_type a, simd_impl<Impl> b) {\
+    return simd_impl<Impl>::wrap(Impl::name(Impl::broadcast(a), b.value_));\
+};
+
+#define BINARY_COMPARISON(name)\
```

Comment:
I made a similar comment in `sve.hpp` — these macros should probably be named along the lines of `ARB_BINARY_COMPARISON_` etc. so as to avoid collisions and to indicate (with the final '_') that they are for internal use only. They can be undefined after use.

Reply:
Okay